### PR TITLE
test=develop mod_key_log Conditional log output

### PR DESCRIPTION
**Pull request by ChengjinWu**
mod_key_log Conditional log output

The conditions are as follows：
                ses_vip_range("10.0.0.1", "10.0.0.10")
                ses_sip_range("10.0.0.1", "10.0.0.10")
                ses_tls_sni_in("example.com|example.org")

### CODE REVIEW

**Comment by iyangsj** [comment]
@ChengjinWu Thanks for your contribution! 
Please visit https://cla-assistant.io/bfenetworks/bfe?pullRequest=580 and sign in with github to agree the CLA for bfe.

**Comment by ChengjinWu** [comment]
> @ChengjinWu Thanks for your contribution!
> Please visit https://cla-assistant.io/bfenetworks/bfe?pullRequest=580 and sign in with github to agree the CLA for bfe.

done @iyangsj

File: conf/mod_key_log/key_log.data [hunk 1]
```diff
@@ -0,0 +1,10 @@
+{
+    "Version": "init version",
+    "Config": {
```
Comment (iyangsj) [inline conf/mod_key_log/key_log.data:3]: 建议默认配置改为：

1. 仅包含 global规则，针对所有流量生效

File: bfe_modules/mod_key_log/mod_key_log.go [hunk 2]
```diff
@@ -52,20 +58,48 @@ func (m *ModuleKeyLog) Name() string {
 }
 
 func (m *ModuleKeyLog) logTlsKey(session *bfe_basic.Session) int {
+	isNeedKeyLog := m.isNeedKeyLog(session)
+	if !isNeedKeyLog {
+		return bfe_module.BfeHandlerGoOn
```
Comment (iyangsj) [inline bfe_modules/mod_key_log/mod_key_log.go:63]: line 61 ~ line62 直接合并为：

if m.isNeedKeyLog()

File: bfe_modules/mod_key_log/mod_key_log.go [hunk 3]
```diff
@@ -52,20 +58,48 @@ func (m *ModuleKeyLog) Name() string {
...
 }
 
+// isNeedKeyLog Determine if you need to print the key log
+func (m *ModuleKeyLog) isNeedKeyLog(session *bfe_basic.Session) bool {
+	rules, ok := m.ruleTable.Search(session.Product)
+	if !ok {
+		rules, ok = m.ruleTable.Search(bfe_basic.GlobalProduct)
+	}
+	if !ok {
+		return true
```
Comment (iyangsj) [inline bfe_modules/mod_key_log/mod_key_log.go:84]: 这里应是  return false

File: bfe_modules/mod_key_log/mod_key_log.go [hunk 4]
```diff
@@ -52,20 +58,48 @@ func (m *ModuleKeyLog) Name() string {
...
+	}
+	flag := false
+	for _, rule := range *rules {
+		// rule condition is satisfied ?
+		if rule.Cond.Match(req) {
+			// do actions of the rule
+			// todo
+
+			// finish key_log rules process
+			flag = true
```
Comment (iyangsj) [inline bfe_modules/mod_key_log/mod_key_log.go:97]: line 93~ line 97 修改为 return true

File: bfe_modules/mod_key_log/mod_key_log.go [hunk 5]
```diff
@@ -52,20 +58,48 @@ func (m *ModuleKeyLog) Name() string {
...
+		}
+	}
+	return flag
```
Comment (iyangsj) [inline bfe_modules/mod_key_log/mod_key_log.go:100]: 这里修改为return false

File: bfe_modules/mod_key_log/mod_key_log.go [hunk 6]
```diff
@@ -91,5 +131,33 @@ func (m *ModuleKeyLog) Init(cbs *bfe_module.BfeCallbacks, whs *web_monitor.WebHa
...
+	// get file path
+	path := query.Get("path")
+	if path == "" {
+		// use default
+		path = m.dataConfigPath
+	}
+
+	// load from config file
+	conf, err := keyLogConfLoad(path)
+
```
Comment (iyangsj) [inline bfe_modules/mod_key_log/mod_key_log.go:153]: 删除 line 153

File: bfe_modules/mod_key_log/key_log_conf_load.go [hunk 7]
```diff
@@ -0,0 +1,177 @@
+// Copyright (c) 2019 Baidu, Inc.
```
Comment (iyangsj) [inline bfe_modules/mod_key_log/key_log_conf_load.go:1]: line 1 修改为：
// Copyright (c) 2019 The BFE Authors.

其它新增文件同

File: bfe_modules/mod_key_log/key_log_conf_load.go [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+}
+
+// keyLogConfLoad loads config of key_log from file.
+func keyLogConfLoad(filename string) (keyLogConf, error) {
+	var conf keyLogConf
+	var err error
+
+	/* open the file    */
+	file, err1 := os.Open(filename)
+
```
Comment (iyangsj) [inline bfe_modules/mod_key_log/key_log_conf_load.go:142]: 删除line 142空行

File: bfe_modules/mod_key_log/key_log_conf_load.go [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+	if err1 != nil {
+		return conf, err1
+	}
+
```
Comment (iyangsj) [inline bfe_modules/mod_key_log/key_log_conf_load.go:146]: line 152删除，这里添加 defer file.Close()

File: bfe_modules/mod_key_log/key_log_conf_load.go [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+	/* decode the file  */
+	decoder := json.NewDecoder(file)
+
+	var config KeyLogConfFile
+	err = decoder.Decode(&config)
+	file.Close()
+
```
Comment (iyangsj) [inline bfe_modules/mod_key_log/key_log_conf_load.go:153]: 删除line 153

File: bfe_modules/mod_key_log/key_log_conf_load.go [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+			return ruleList, err
+		}
+		*ruleList = append(*ruleList, rule)
+	}
+
+	return ruleList, nil
+}
+
+// keyLogConfLoad loads config of key_log from file.
+func keyLogConfLoad(filename string) (keyLogConf, error) {
```
Comment (iyangsj) [inline bfe_modules/mod_key_log/key_log_conf_load.go:136]: 注释统一使用 // 风格

**Comment by ChengjinWu** [comment]
@iyangsj 上面提到的问题，我都改完了🙂

**Pull request merged**